File: leetcode/Dynamic Programming/combination_sum_IV.erl
%% https://leetcode.com/problems/combination-sum-iv/

-module(combination_sum_IV).

-compile(export_all).

-spec combination_sum4(Nums :: [integer()], Target :: integer()) -> integer().
combination_sum4(Nums, Target) ->
    #{Target := Res} = inner(
        Nums,
        Target
    ),
    Res.

inner(Nums, Target) ->
    inner(1, Target, Nums, #{0 => 1}).

inner(CurrTarget, MaxTarget, _Nums, Map) when CurrTarget > MaxTarget -> Map;
inner(Target, MaxTarget, Nums, Map) ->
    Fun = fun(Num, Acc) -> Acc + maps:get(Target - Num, Map, 0) end,
    inner(Target+1, MaxTarget, Nums, Map#{Target => lists:foldl(Fun, 0, Nums)}).
    %% Map[Target] = 
    %%     iterate through Nums 
    %%     Acc += Map.get(Target - Num, 0)  
    %%      

File: leetcode/Dynamic Programming/combination_sum_IV.cpp
// https://leetcode.com/problems/combination-sum-iv/

class Solution {
public:
    int combinationSum4(vector<int>& nums, int target) {
        vector<unsigned int> dp (target + 1, 0);
        for (int i = 0; i <= target; i++) {
            for (int j = 0; j < nums.size(); j++) {
                if (dp[i] == 4611686018427387904) {
                    continue;
                }
                if (nums.at(j) > i) {
                    continue;
                }
                else if (nums.at(j) == i) {
                    dp[i] += 1;
                }
                else {
                    dp[i] += dp[i - nums[j]];
                }
            }
            cout << i << " "<< dp[i] << endl;
            // cout << dp[i] << endl;
        }
        return (int)(dp[target]);
    }
};
